syncer/streamer_controller.go

```diff
@@ -0,0 +1,341 @@
+	// binlog type is remote: dm-worker has more than one sub task, so need generate random server id
+	randomServerID, err := utils.GetRandomServerID(tctx.Context(), c.fromDB.BaseDB.DB)
+	if err != nil {
+		// should never happened unless the master has too many slave
+		return terror.Annotate(err, "fail to get random server id for streamer controller")
+	}
+	c.syncCfg.ServerID = randomServerID
+
+	if c.initBinlogType == LocalBinlog && c.meetError {
+		// meetError is true means meets error when get binlog event, in this case use remote binlog as default
+		if !uuidSameWithUpstream {
+			// if the binlog position's uuid is different from the upstream, can not switch to remote binlog
+			tctx.L().Warn("may switch master in upstream, so can not switch local to remote")
+		} else {
+			c.currentBinlogType = RemoteBinlog
+			tctx.L().Warn("meet error when read from local binlog, will switch to remote binlog")
+		}
+	}
+
+	if c.currentBinlogType == RemoteBinlog {
+		c.streamerProducer = &remoteBinlogReader{replication.NewBinlogSyncer(c.syncCfg), tctx, false}
+	} else {
+		c.streamerProducer = &localBinlogReader{streamer.NewBinlogReader(tctx, &streamer.BinlogReaderConfig{RelayDir: c.localBinlogDir, Timezone: c.timezone})}
+	}
+
+	c.streamer, err = c.streamerProducer.generateStreamer(pos)
+	return err
+}
+
+// RedirectStreamer redirects the streamer's begin position or gtid
+func (c *StreamerController) RedirectStreamer(tctx *tcontext.Context, pos mysql.Position) error {
+	tctx.L().Info("reset global streamer", zap.Stringer("position", pos))
+	return c.ResetReplicationSyncer(tctx, pos)
+}
+
+// GetEvent returns binlog event
+func (c *StreamerController) GetEvent(tctx *tcontext.Context) (event *replication.BinlogEvent, err error) {
+	ctx, cancel := context.WithTimeout(tctx.Context(), eventTimeout)
+	failpoint.Inject("SyncerEventTimeout", func(val failpoint.Value) {
+		if seconds, ok := val.(int); ok {
+			cancel()
+			ctx, cancel = context.WithTimeout(tctx.Context(), time.Duration(seconds)*time.Second)
+			tctx.L().Info("set fetch binlog event timeout", zap.String("failpoint", "SyncerEventTimeout"), zap.Int("value", seconds))
+		}
+	})
+
+	event, err = c.streamer.GetEvent(ctx)
+	cancel()
+	if err == nil {
+		switch ev := event.Event.(type) {
+		case *replication.RotateEvent:
+			// if is local binlog, binlog's name contain uuid information, need to save it
+			// if is remote binlog, need to add uuid information in binlog's name
+			if !c.setUUIDIfExists(string(ev.NextLogName)) {
+				if len(c.uuidSuffix) != 0 {
+					filename, err := binlog.ParseFilename(string(ev.NextLogName))
+					if err != nil {
+						return nil, terror.Annotate(err, "fail to parse binlog file name from rotate event")
+					}
+					ev.NextLogName = []byte(binlog.ConstructFilenameWithUUIDSuffix(filename, c.uuidSuffix))
+					event.Event = ev
+				}
+			}
+		default:
+		}
+		return event, nil
+	}
+
+	if err != context.Canceled {
+		c.meetError = true
```

Comment:
yes

Reply:
There is no place to set `meetError` to `false`, so after `context.DeadlineExceeded` happened (like no write operation in upstream for a long time), then it may switch to `remote` for later `canceled` error?

Reply:
seems it is better to ignore set `meetError` to true when err is `DeadlineExceeded`, update in f99706a